library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileInputOperator.java

```diff
@@ -569,9 +601,18 @@ protected void replay(long windowId)
       Map<Integer, Object> recoveryDataPerOperator = windowDataManager.retrieveAllPartitions(windowId);
 
       for (Object recovery : recoveryDataPerOperator.values()) {
-        @SuppressWarnings("unchecked")
-        LinkedList<RecoveryEntry> recoveryData = (LinkedList<RecoveryEntry>)recovery;
 
+        WindowRecoveryEntry windowRecoveryEntry = (WindowRecoveryEntry)recovery;
+        List<RecoveryEntry> recoveryData = windowRecoveryEntry.fileRecoveryInfo;
+
+        if (windowRecoveryEntry.startBatchEmitted) {
+          emitStartBatchControlTuple();
```

Comment:
It seems something is wrong here. Do we want to emit multiple start and end batch tuples in the replay code?
This might mess up the downstream operators. Remember that all this is done in a single window.
The right thing to do here would be to do a replay but also at the same speed. We do have the files and offsets that are completed in each window. Perhaps we can keep repeating this until all data is replayed. Note we do not increment the windowDataManager state during this time. Once all replayed data is emitted, then in the next window we can start with reading new data.
@PramodSSImmaneni @vrozov thoughts?